QueryService: Use types from sdk

ryantxu opened the pull request:
Extracted from https://github.com/grafana/grafana/pull/82391

This can be tested manually be setting `kubernetesQueryServiceRewrite` feature flag -- this will redirect all requests to /ds/query to /apis/query.grafana.app/v0alpha/namespaces/default/query

ryantxu committed:
extract from bigger PR

ryantxu committed:
running

ryantxu committed:
remove runner implementations

ryantxu committed:
remove some files

ryantxu committed:
improve comments

ryantxu committed:
consolidate files

ryantxu committed:
consolidate files

ryantxu commented on pkg/services/datasources/service/legacy.go:
```diff
@@ -0,0 +1,90 @@
+type LegacyDataSourceLookup interface {
+	// Find the UID from either the name or internal id
+	// NOTE the orgID will be fetched from the context
+	GetDataSourceFromDeprecatedFields(ctx context.Context, name string, id int64) (*data.DataSourceRef, error)
```
@marefr -- I know we could use the `DataSourceRetriever` interface, but with this approach we can isolate all the actual configs.  Usage of this also does not need to clear a cache since nothing new should produce dashboards that need to lookup by name or id.

ryantxu committed:
Merge remote-tracking branch 'origin/main' into query-service-use-sdk-models

ryantxu commented on pkg/apis/query/v0alpha1/datasource.go:
```diff
@@ -3,26 +3,10 @@ package v0alpha1
 )
 
-// The query runner interface
-type QueryRunner interface {
```
this is replaced with the client interface in the SDK

ryantxu commented on pkg/registry/apis/query/metrics.go:
```diff
@@ -0,0 +1,48 @@
+	expressionsQuerySummary *prometheus.SummaryVec
+}
+
+func newMetrics(reg prometheus.Registerer) *metrics {
```
this is just copied from expressions

toddtreece commented on pkg/registry/apis/query/query.go:
```diff
@@ -150,8 +246,67 @@ func (b *QueryAPIBuilder) executeConcurrentQueries(ctx context.Context, requests
+					if err != nil {
+						res.Error = err
+					}
+					vars[expression.RefID] = res
```
what's the difference between `expression.RefID` and `refId`?

ryantxu replied:
a mistake 🤦🏻 -- fixed.  expression.RefID is the expressions ID, but "refId" is the id for the result it requires.  The map should be full of the results id

toddtreece commented on pkg/registry/apis/query/register.go:
```diff
@@ -115,6 +133,7 @@ func (b *QueryAPIBuilder) GetAPIGroupInfo(
 
 	storage := map[string]rest.Storage{}
 	storage[plugins.resourceInfo.StoragePath()] = plugins
+	storage["query"] = &subQueryREST{builder: b}
```
does this work? i'd assume this would need to be something like:
```suggestion
	storage[plugins.resourceInfo.StoragePath()+"/query"] = &subQueryREST{builder: b}
```

ryantxu replied:
with "query" this gets attached to:
<img width="881" alt="image" src="https://github.com/grafana/grafana/assets/705951/9318a85d-dc9f-4419-85e2-206dabeee85a">

It is currently a Creator -- but I think we should change it back to a raw http handler because there is enough weird about it that I do not like (the response codes, logging, etc)

toddtreece replied:
ah ok, i was confused by the name `subQueryREST`

toddtreece commented on pkg/registry/apis/query/client/testdata.go:
```diff
@@ -1,59 +1,57 @@
+				break
+			}
+		}
+	} else {
```
this logic is used in a couple places. should it be a function?

ryantxu replied:
done

toddtreece commented on pkg/registry/apis/query/parser.go:
```diff
@@ -1,83 +1,209 @@
-				To:   raw.To,
+		// Process each query
+		if expr.IsDataSource(ds.UID) {
+			raw, err := json.Marshal(q)
```
could you add a comment here explaining why it's going from object -> bytes -> iter -> exp object?

ryantxu committed:
Merge remote-tracking branch 'origin/main' into query-service-use-sdk-models

ryantxu commented on pkg/registry/apis/query/parser.go:
```diff
@@ -1,83 +1,209 @@
+		input.From = "now-6h"
+	}
+	if input.To == "" {
+		input.To = ""
```
good catch -- merge error

ryantxu committed:
review feedback

toddtreece commented on pkg/services/datasources/service/legacy.go:
```diff
@@ -0,0 +1,90 @@
+type cachingLegacyDataSourceLookup struct {
+	retriever DataSourceRetriever
+	cache     map[string]cachedValue
+	cacheMu   sync.Mutex
```
should this be a `sync.RWMutex`?

toddtreece commented on pkg/registry/apis/query/query.go:
```diff
@@ -2,92 +2,185 @@ package query
 	}
-	return b.runner.ExecuteQueryData(ctx, gv, req.uid, req.query)
+
+	// headers?
```
is this an important question to answer before this merges?

ryantxu replied:
Not before this merge -- it will be important for how we get auth headers from alerting client code passed along to each of the steps; but I think it is too much to get sorted in this one.

ryantxu committed:
expression support

ryantxu committed:
add expression in test example

ryantxu commented:
@toddtreece -- since you looked at this before I made two major changes:

1. changed from using Creator to a raw http handler.   This is not great, but I am not sure how else to solve the issues that we need to return a 200, not a 201 for a query request.  There is still more work to make the openapi spec accurate/better, but i think that can come in a later PR
2. updated the test to include an expression -- very simple: $A + 2, but it does exercise the issues.

To manually test the behavior, set feature toggle: `kubernetesQueryServiceRewrite` and that sends all requests to this endpiont

toddtreece commented on pkg/registry/apis/query/parser_test.go:
```diff
@@ -0,0 +1,131 @@
+		require.Empty(t, split.Requests)
+	})
+
+	t.Run("applies default timerage", func(t *testing.T) {
```
```suggestion
	t.Run("applies default time😡", func(t *testing.T) {
```

ryantxu replied:
🤣 🤦🏻

toddtreece commented on pkg/registry/apis/query/testdata/with-expressions.json:
```diff
@@ -0,0 +1,38 @@
+    ]
+  },
+  "expect": {},
+  "error": "missing type"
```
is this expected in these test files? `"error": "missing type"`

toddtreece reviewed (approved):
it worked for me locally 🎉

tested with:
```
[feature_toggles]
kubernetesQueryServiceRewrite = true
grafanaAPIServerWithExperimentalAPIs = true
```
 
at this point this PR is getting tough to review because of its size. it seems safe to merge and iterate on in future PRs

ryantxu committed:
fix typos and tests
